File: #2CTinGĐ1619/12CTin/HSG-TPQG/HSG_12_TP/14/jump.pas
uses math;
var x,y: array [0..10000] of int64;
    n,h,i,u,r,first,rear: longint;
    queue, rank: array [0..10000] of longint;
    fre: array [0..10000] of boolean;

procedure push(i,r: longint);
begin
    inc(rear);
    queue[rear]:=i;
    fre[i]:=true;
    rank[i]:=r;
end;

function tinh(i,u: longint): longint;
begin
        exit(trunc(sqrt(sqr(x[i] - x[u]) + sqr(y[i] - y[u]))));
end;

BEGIN
        assign(input,'jump.inp'); reset(input);
        assign(output,'jump.out'); rewrite(output);

        read(h,n);
        for i:=1 to n do read(x[i], y[i]);

        rear:=-1;
        for i:=1 to n do if y[i] <= 1000 then push(i,1);
        repeat
                u:=queue[first]; inc(first);
                r:=rank[u];
                if (h - y[u] <= 1000) then begin
                        write(r); exit;
                end;
                for i:=1 to n do
                        if not fre[i] and (tinh(i,u) <= 1000) then push(i,r + 1);
        until first > rear;
END.

File: #2CTinGĐ1619/12CTin/HSG-TPQG/HSG_12_TP/14/jump.cpp
#include <bits/stdc++.h>
using namespace std;

int n, h, qu[10001], a[10001], b[10001], bac[10001];
bool fre[10001];

int tinh(int u, int v, int x, int z) {
	return 1000.0 >= sqrt(1.0 * (u - x) * (u - x) + 1.0 * (v - z) * (v - z));
}

int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	freopen("jump.inp","r",stdin);
	freopen("jump.out","w",stdout);

	cin >> h >> n;
	for (int i = 1; i <= n; i++) cin >> a[i] >> b[i];

	int fi = 0, se = -1;
	for (int i = 1; i <= n; i++) if (b[i] <= 1000) qu[++se] = i, bac[i] = 1, fre[i] = true;

	while (fi <= se) { 
		int u = a[qu[fi]], v = b[qu[fi]], note = bac[qu[fi]] + 1;

		for (int j = 1; j <= n; j++) 
			if (qu[fi] != j && !fre[j]) {
				if (tinh(u,v,a[j],b[j])) qu[++se] = j, bac[j] = note, fre[j] = true;				
				if (tinh(u,v,u,h)) {
					cout << note - 1; 
					return 0;
				} 
			}

		fi++;
	}
}